File: himark/parser/phase0.py
"""Phase 0: Split a raw HMK statement into its ordered list of steps.

A statement is a `=>` chain: `pattern => pattern => ... => template`. There is a
single arrow — `=>`. The chain produces one **branch** per match of the first
pattern; each branch is transformed independently and rendered either as a list
(the rendered matches) or spliced back into the source (in-place transform). The
two renderings come from the same branches, so there is no separate replace arrow.
"""


def split_statement(text: str) -> list[str]:
    """Split `'P1 => P2 => ... => T'` into its ordered list of step strings.

    Scans for `=>` outside of any construct delimiters to avoid false splits.
    """
    steps: list[str] = []
    remaining = text
    while True:
        idx = _find_arrow(remaining)
        if idx is None:
            steps.append(remaining.strip())
            break
        steps.append(remaining[:idx].strip())
        remaining = remaining[idx + 2 :]
    return steps


def _find_arrow(text: str) -> int | None:
    """Index of the first top-level `=>`, or None.

    An arrow is recognised only at top level — outside every `{…}`, `[…]`, and
    `"…"` (ANTLR friendly): inside a quoted template `=>` is literal text and never
    splits, so a template may contain `=>` freely with no escape. A single `<` or
    `>` is plain text (a template's `<strong>` never reads as an arrow), and a
    backslash-escaped character is never a delimiter.
    """
    depth = 0
    inq = False
    i = 0
    while i < len(text):
        ch = text[i]
        if ch == "\\":
            i += 2
            continue
        if ch == '"':
            inq = not inq
        elif inq:
            pass  # inside a quoted template — braces and `=>` are literal
        elif ch == "=" and text[i + 1 : i + 2] == ">" and depth == 0:
            return i
        elif ch in ("[", "{"):
            depth += 1
        elif ch in ("]", "}"):
            depth = max(0, depth - 1)
        i += 1
    return None

File: rust/src/parser/phase0.rs
/// Phase 0: split a raw HMK statement on top-level `=>` arrows.
/// Mirrors himark/parser/phase0.py.

pub fn split_statement(text: &str) -> Vec<String> {
    let mut steps: Vec<String> = Vec::new();
    let mut remaining = text;
    loop {
        match find_arrow(remaining) {
            None => {
                steps.push(remaining.trim().to_string());
                break;
            }
            Some(idx) => {
                steps.push(remaining[..idx].trim().to_string());
                remaining = &remaining[idx + 2..];
            }
        }
    }
    steps
}

/// Returns the byte index of the first top-level `=>`, or `None`.
fn find_arrow(text: &str) -> Option<usize> {
    let bytes = text.as_bytes();
    let n = text.len();
    let mut depth = 0i32;
    let mut i = 0usize;
    while i < n {
        let b = bytes[i];
        if b == b'\\' {
            i += 2;
            continue;
        }
        if b == b'=' && i + 1 < n && bytes[i + 1] == b'>' && depth == 0 {
            return Some(i);
        }
        if b == b'[' || b == b'{' {
            depth += 1;
        } else if b == b']' || b == b'}' {
            depth = depth.saturating_sub(1);
        }
        i += 1;
    }
    None
}
